Update documentation for leak tracker. (#26)
dart-lang/leak_tracker: README.md

```

2. Run the application in debug mode and watch for a leak related warnings.
If you see a warning, open the link to investigate the leaks.

TODO(polina-c): implement the link and add example of the warning.

### Flutter tests

Wrap your tests with `withLeakTracking` to setup automated leak verification:

```dart
test('...', () async {
  final leaks = await withLeakTracking(
    () async {
      ...
    },
  );

>   expect(leaks, leakFree);

Reviewer: What is `leaksFree`? I can't seem to find a reference to it in the repo

Author: Will show up in the next PR

2. Run the application in debug mode and watch for a leak related warnings. If you see a warning, open the link to investigate the leaks.

TODO(polina-c): implement the link and add example of the warning.

### Flutter tests

Wrap your tests with `withLeakTracking` to setup automated leak verification:

```dart
test('...', () async {
  final leaks = await withLeakTracking(
    () async {
      ...
    },
    throwOnLeaks: false,
  );

  expect(leaks.total, 0);
});
```

## Leak tracking concepts

> Before reading about leak tracking, understand [Dart memory concepts](https://docs.flutter.dev/development/tools/devtools/memory#basic-memory-concepts).

Author: Normally 'Quick start' means the person does not want to learn anything, they just wants to make it working. So, I would leave it here. 

But, tech writers may agree with you, so leaving the comment unresolved.

### Addressed leak types

The leak tracker can catch only certain types of leaks, in particular, related to timing of disposal and garbage collection.

The tool assumes that, with proper memory management,
an object's disposal and garbage collection should happen sequentially,
close to each other.

> By monitoring disposal and GC events, the tool detects different types of leaks:

Reviewer: ```suggestion
By monitoring disposal and Garbage Collect events, the tool detects different types of leaks:
```
Since this is an instructive document, it may help to spell out "Garbage Collect" for GC and "Garbage Collected" from GCed.
We don't have any space constraints here, so I think clarity takes precedence over brevity

> **Not disposed, but GCed (not-disposed)**: a disposable object was GCed,

Reviewer: It may be good to split the bodies of these enties into "Description" and "Fix" bullets, so that it is easy to visually parse the sections and their value.

without being disposed first. This means that the object's disposable content
is using memory after the object is no longer needed.

> To fix the leak, invoke `dispose()` to free up the memory.

Reviewer: what should the reader call dispose on? it might be helpful to mention that here

**Disposed, but not GCed (not-GCed)**: an object was disposed,
but not GCed after certain number of GC events. This means that
a reference to the object is preventing it from being
garbage collected after it's no longer needed.
To fix the leak, after disposal assign all reachable references
of the object to null:

```
myField.dispose();
myField = null;
```

**Disposed and GCed late (GCed-late)**: an object was disposed and then GCed,
but GC happened later than expected. This means the retaining path was

> holding the object in memory for some period, but then disappeared.

Reviewer: Is there a hint or fix we can lead them to here?

> **Disposed and GCed late (GCed-late)**: an object was disposed and then GCed,
> but GC happened later than expected. This means the retaining path was
> holding the object in memory for some period, but then disappeared.
> 
> **Disposed, but not GCed, without path (not-GCed-without-path)**: an object
> was disposed and not GCed when expected, but retaining path is not detected,
> that means that the object will be most likely GCed in the next GC cycle,
> and the leak will convert to **GCed-late** leak.

Reviewer: ```suggestion
- **Disposed and GCed late (GCed-late)**:
    - an object was disposed and then GCed,
       but GC happened later than expected. This means the retaining path was
       holding the object in memory for some period, but then disappeared.

- **Disposed, but not GCed, without path (not-GCed-without-path)**:
    - an object
       was disposed and not GCed when expected, but retaining path is not detected,
       that means that the object will be most likely GCed in the next GC cycle,
       and the leak will convert to **GCed-late** leak.
```

> ### Culprits and victims

Reviewer: This terminology may end up being a bit too harsh. It may be worth asking the tech writers if they have different analogies they would prefer using here.

Author: Tech writers and UXR already reviewed it and they loved it, because it is extremely illustrative :)

If you have a set of not-GCed objects, some of them (victims)
might not be GC-ed because they are held by others (culprits).
Normally, to fix the leaks, you need to only fix the culprits.

**Victim**: a leaked object, for which the tool could find another
leaked object that, if fixed, would also fix the first leak.

**Culprit**: a leaked object that is not detected to be the victim
of another object.

The tool detects which leaked objects are culprits, so you know where to focus.

For example, out of four not-GCed leaks on the following diagram,
only one is the culprit, because, when the object is fixed
and GCed, the victims it referenced will be also GCed:

```mermaid

   flowchart TD;
      l1[leak1\nculprit]
      l2[leak2\nvictim]
      l3[leak3\nvictim]
      l4[leak4\nvictim]
      l1-->l2;
      l1-->l3;
      l2-->l4;
      l3-->l4;
```

## Limitations

### By tracked classes

> The leak traker will catch leaks only for instrumented onjects (See [concepts](#leak-tracking-concepts) for details).

Reviewer: ```suggestion
The leak tracker will catch leaks only for instrumented objects (See [concepts](#leak-tracking-concepts) for details).
```

> However, the good news are:

Reviewer: ```suggestion
However, the good news is:
```

1. Most of disposable Flutter Framework classes are already instrumented. So, Widget related

> leaks in your Flutter application will be cought without additional effort.

Reviewer: ```suggestion
leaks in your Flutter application will be caught without additional effort.
```

2. If a leak involves at least one instrumented object, the leak will be caught and all
other objects, even non-instrumented, will stop leaking as well.

See [the instrumentation guidance](#instrument-your-code).

> 

Reviewer: It might be a good idea to add a section mentioning how to report a bug if they notice that a framework entity has a leak

Author: Thanks. Updated the sentence to show it is most likely in-app problem.

### By build mode

The leak tracker availability differs by build modes. See [Dart build modes](https://github.com/dart-lang/site-www/issues/4436) or [Flutter build modes](https://docs.flutter.dev/testing/build-modes).

**Dart development and Flutter debug**

Leak tracking is fully available.

**Flutter profile**

Leak tracking is available, but MemoryAllocations that listens to Flutter instrumented objects,
should be [turned on](https://github.com/flutter/flutter/blob/15af81782e19ebe7273872f8b07ac71df4e749f2/packages/flutter/lib/src/foundation/memory_allocations.dart#L13) if you want to track Flutter Framework objects.

**Dart productive and Flutter release**

Leak tracking is disabled.

NOTE: If you are interested in enabling leak tracking for release mode, please, comment [here](https://github.com/dart-lang/leak_tracker/issues/25).

## Instrument your code

If you want to catch leaks for objects outside of Flutter Framework,
(that are already instrumented) you need to instrument them.

For each tracked object the library should get two signals from your code:

> (1) the object is created and (2) the object is not in use. It is most convenient to give first signal in constractor and the second one in the method `dispose`:

Reviewer: ```suggestion
(1) the object is created and (2) the object is not in use. It is most convenient to give the first signal in the constructor and the second one in the `dispose` method:
```

    - **Fix**: invoke `dispose()` for the object to free up the memory.

- **Disposed, but not GCed (not-GCed)**:
    - **Definition**: an object was disposed,
       but not GCed after certain number of GC events. This means that
       a reference to the object is preventing it from being
       garbage collected after it's no longer needed.

    - **Fix**: To fix the leak, assign all reachable references
       of the object to null after disposal:

        ```
        myField.dispose();
        myField = null;
        ```

- **Disposed and GCed late (GCed-late)**:
    - **Definition**: an object was disposed and then GCed,
       but GC happened later than expected. This means the retaining path was
       holding the object in memory for some period, but then disappeared.

    - **Fix**: the same as for **not-GCed**

- **Disposed, but not GCed, without path (not-GCed-without-path)**:
    - **Definition**: an object
       was disposed and not GCed when expected, but retaining path is not detected,
       that means that the object will be most likely GCed in the next GC cycle,
       and the leak will convert to **GCed-late** leak.

>     - **Fix**: please, [create issue](https://github.com/dart-lang/leak_tracker/issues) if you see this type of leaks, as it means something is wrong with the tool.

Reviewer 2: **issue:** This information would be better presented as a table. I would replace it with the following:
```suggestion
| Memory Leak Type          | Disposed | Garbage Collected | Explanation                                                                                                                                                                                                                                   | Fix                                                                                                                              |   |
|---------------------------|:--------:|:-----------------:|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------|---|
| **not-disposed**          |    No    |        Yes        | The object's content continues to use memory after the program no longer needs the object.                                                                                                                                                    | Call `dispose()` to free up the memory.                                                                                          |   |
| **not-GCed**              |    Yes   |         No        | Memory disposed of an object, but did not free the memory that object occupied after certain number of GC events. A reference to the object prevents removing it from memory after the program no longer needs it to be in memory.            | Assign all reachable references of the object to `null` after disposal.  ```dart myField.dispose(); myField = null; ```          |   |
| **GCed-late**             |    Yes   |   Yes, but late   | Memory disposed of an object, then cleared the object from memory later than expected. The retaining path continued to hold the object in memory for some period, but then disappeared.                                                       | Assign all reachable references of the object to `null` after disposal.  ```dart myField.dispose(); myField = null; ```          |   |
| **not-GCed-without-path** |    Yes   |  No, without path | Memory disposed of an object and did not clear the object from memory when expected, but memory cannot detect the object's retaining path. The object should be cleared in the next GC cycle. The leak then converts to a **GCed-late** leak. | [Create an issue](https://github.com/dart-lang/leak_tracker/issues). If you see this type of leak, the tool might have an issue. |   |
```

Author: I love how the table look, but the table format makes it very hard to edit and review the page. So, leaving it bulleted list for the sake of maintainability.

However, the good news is:

> 1. Most of disposable Flutter Framework classes are already instrumented. So, if
> your Flutter application manages widgets in a way that creates leaks, the leaks
> will be caught without additional effort.

Reviewer 2: **issue:** This reads with some difficulty. I would recommend rephrasing somewhat.
```suggestion
1. Most disposable Flutter Framework classes include instrumentation.
If how your Flutter app manages widgets results in leaks, Flutter will catch them.
```
